primitiv/c/shape.cc

```diff
@@ -5,101 +5,63 @@
-                                                  primitiv_Status *status) {
-  SAFE_RETURN(primitiv_Shape_new_with_dims(dims, n, batch), status, nullptr);
+  try {
+    *shape = to_c(new Shape(std::vector<uint32_t>(dims, dims + n), batch));
+    return ::primitiv_Status::PRIMITIV_OK;
+  } HANDLE_EXCEPTION
 }
 
 void primitiv_Shape_delete(primitiv_Shape *shape) {
   delete to_cc(shape);
 }
-void safe_primitiv_Shape_delete(primitiv_Shape *shape,
-                                primitiv_Status *status) {
-  SAFE_EXPR(primitiv_Shape_delete(shape), status);
-}
 
 uint32_t primitiv_Shape_op_getitem(const primitiv_Shape *shape, uint32_t i) {
   return to_cc(shape)->operator[](i);
 }
-uint32_t safe_primitiv_Shape_op_getitem(const primitiv_Shape *shape,
-                                        uint32_t i,
-                                        primitiv_Status *status) {
-  SAFE_RETURN(primitiv_Shape_op_getitem(shape, i), status, 0);
-}
 
 void primitiv_Shape_dims(const primitiv_Shape *shape, uint32_t *array) {
   std::vector<uint32_t> v = to_cc(shape)->dims();
   std::copy(v.begin(), v.end(), array);
 }
-void safe_primitiv_Shape_dims(const primitiv_Shape *shape,
-                              uint32_t *array,
-                              primitiv_Status *status) {
-  SAFE_EXPR(primitiv_Shape_dims(shape, array), status);
-}
 
 uint32_t primitiv_Shape_depth(const primitiv_Shape *shape) {
   return to_cc(shape)->depth();
 }
-uint32_t safe_primitiv_Shape_depth(const primitiv_Shape *shape,
-                                   primitiv_Status *status) {
-  SAFE_RETURN(primitiv_Shape_depth(shape), status, 0);
-}
 
 uint32_t primitiv_Shape_batch(const primitiv_Shape *shape) {
   return to_cc(shape)->batch();
 }
-uint32_t safe_primitiv_Shape_batch(const primitiv_Shape *shape,
-                                   primitiv_Status *status) {
-  SAFE_RETURN(primitiv_Shape_batch(shape), status, 0);
-}
 
 uint32_t primitiv_Shape_volume(const primitiv_Shape *shape) {
   return to_cc(shape)->volume();
 }
-uint32_t safe_primitiv_Shape_volume(const primitiv_Shape *shape,
-                                    primitiv_Status *status) {
-  SAFE_RETURN(primitiv_Shape_volume(shape), status, 0);
-}
 
 uint32_t primitiv_Shape_lower_volume(const primitiv_Shape *shape,
                                      uint32_t dim) {
   return to_cc(shape)->lower_volume(dim);
 }
-uint32_t safe_primitiv_Shape_lower_volume(const primitiv_Shape *shape,
-                                          uint32_t dim,
-                                          primitiv_Status *status) {
-  SAFE_RETURN(primitiv_Shape_lower_volume(shape, dim), status, 0);
-}
 
 uint32_t primitiv_Shape_size(const primitiv_Shape *shape) {
   return to_cc(shape)->size();
 }
-uint32_t safe_primitiv_Shape_size(const primitiv_Shape *shape,
-                                  primitiv_Status *status) {
-  SAFE_RETURN(primitiv_Shape_size(shape), status, 0);
-}
 
 char *primitiv_Shape_to_string(const primitiv_Shape *shape) {
```

Comment:
In this function, return value `c` maybe doesn't has null-terminated string(`\0`).
I think it's useful to contain it.

Reply:
Although we have to use `strncpy` carefully, `strcpy` copies chars including the null terminator.
See https://linux.die.net/man/3/strcpy .

Reply:
And, this function is dangerous because it returns a memory allocated by internal `new[]` operator. Users can't know which deallocator can be used to free the returned memory.

Reply:
I found `strncpy` in `primitiv_Shape_to_string`. I'll fix it. Thank you.

Reply:
I reimplement `primitiv_Shape_to_string` and `primitiv_Graph_dump` not to allocate memory internally. Instead, I'll provide some way to examine the length of returning characters.